faster_whisper/feature_extractor.py

```diff
@@ -54,61 +77,88 @@ def get_mel_filters(sr, n_fft, n_mels=128):
 
         mel_f = freqs
 
+        # Compute the difference between adjacent mel frequencies
         fdiff = torch.diff(mel_f)
+
+        # Create ramps for lower and upper edges
         ramps = mel_f.view(-1, 1) - fftfreqs.view(1, -1)
 
         lower = -ramps[:-2] / fdiff[:-1].unsqueeze(1)
         upper = ramps[2:] / fdiff[1:].unsqueeze(1)
 
-        # Intersect them with each other and zero, vectorized across all i
+        # Intersect them with each other and zero
         weights = torch.maximum(torch.zeros_like(lower), torch.minimum(lower, upper))
 
         # Slaney-style mel is scaled to be approx constant energy per channel
-        enorm = 2.0 / (mel_f[2 : n_mels + 2] - mel_f[:n_mels])
-        weights *= enorm.unsqueeze(1)
+        enorm = 2.0 / (mel_f[2 : n_mels + 2] - mel_f[:n_mels]).unsqueeze(1)
+        weights *= enorm
 
         return weights
 
-    def __call__(self, waveform, padding=True, chunk_length=None, to_cpu=False):
-        """
-        Compute the log-Mel spectrogram of the provided audio.
+    def __call__(
+        self,
+        waveform: torch.Tensor,
+        padding: bool = True,
+        chunk_length: int = None,
+        to_cpu: bool = False,
+    ) -> torch.Tensor:
         """
+        Compute the log-Mel spectrogram of the provided audio waveform.
 
+        Args:
+            waveform (torch.Tensor): Input audio waveform tensor of shape (..., n_samples).
+            padding (bool): Whether to pad the waveform to the required number of samples.
+            chunk_length (int, optional): Length of audio chunks in seconds. Overrides initialization if provided.
+            to_cpu (bool): Whether to move the output spectrogram to CPU.
+
+        Returns:
+            torch.Tensor: Log-Mel spectrogram tensor.
+        """
+        # Use local variables instead of modifying instance attributes
         if chunk_length is not None:
-            self.n_samples = chunk_length * self.sampling_rate
-            self.nb_max_frames = self.n_samples // self.hop_length
+            n_samples = chunk_length * self.sampling_rate
+            nb_max_frames = n_samples // self.hop_length
+        else:
+            n_samples = self.n_samples
+            nb_max_frames = self.nb_max_frames
 
+        # Ensure waveform is float32
         if waveform.dtype is not torch.float32:
             waveform = waveform.to(torch.float32)
 
-        waveform = (
-            waveform.to(self.device)
-            if self.device == "cuda" and not waveform.is_cuda
-            else waveform
-        )
+        # Move waveform to the target device if necessary
+        if self.device == "cuda" and not waveform.is_cuda:
+            waveform = waveform.to(self.device)
 
+        # Apply padding if required
         if padding:
-            waveform = torch.nn.functional.pad(waveform, (0, self.n_samples))
-
-        window = torch.hann_window(self.n_fft).to(waveform.device)
```

Comment:
Why was the `hann_window` deleted?

Reply:
I'll take a look...

Reply:
This is actually an optimization. In the old version, a new Hann window was being created and moved to the device every time __call__ was executed. The new version creates it once during initialization and caches it as an instance variable (self.window).